plugins/builtin/target/aws-asg/plugin/plugin.go

```diff
@@ -0,0 +1,181 @@
+	configKeyDrainDeadline = "drain_deadline"
+
+	// configValues are the default values used when a configuration key is not
+	// supplied by the operator that are specific to the plugin.
+	configValueRegionDefault = "us-east-1"
+)
+
+var (
+	PluginConfig = &plugins.InternalPluginConfig{
+		Factory: func(l hclog.Logger) interface{} { return NewAWSASGPlugin(l) },
+	}
+
+	pluginInfo = &base.PluginInfo{
+		Name:       pluginName,
+		PluginType: plugins.PluginTypeTarget,
+	}
+)
+
+// Assert that TargetPlugin meets the target.Target interface.
+var _ target.Target = (*TargetPlugin)(nil)
+
+// TargetPlugin is the AWS ASG implementation of the target.Target interface.
+type TargetPlugin struct {
+	config       map[string]string
+	logger       hclog.Logger
+	asg          *autoscaling.Client
+	ec2          *ec2.Client
+	scaleInUtils *scaleutils.ScaleIn
+}
+
+// NewAWSASGPlugin returns the AWS ASG implementation of the target.Target
+// interface.
+func NewAWSASGPlugin(log hclog.Logger) *TargetPlugin {
+	return &TargetPlugin{
+		logger: log,
+	}
+}
+
+// SetConfig satisfies the SetConfig function on the base.Plugin interface.
+func (t *TargetPlugin) SetConfig(config map[string]string) error {
+
+	t.config = config
+
+	if err := t.setupAWSClients(config); err != nil {
+		return err
+	}
+
+	utils, err := scaleutils.NewScaleInUtils(nomad.ConfigFromMap(config), t.logger)
+	if err != nil {
+		return err
+	}
+	t.scaleInUtils = utils
+
+	return nil
+}
+
+// PluginInfo satisfies the PluginInfo function on the base.Plugin interface.
+func (t *TargetPlugin) PluginInfo() (*base.PluginInfo, error) {
+	return pluginInfo, nil
+}
+
+// Scale satisfies the Scale function on the target.Target interface.
+func (t *TargetPlugin) Scale(action strategy.Action, config map[string]string) error {
+
+	// We cannot scale an ASG without knowing the ASG name.
+	asgName, ok := config[configKeyASGName]
+	if !ok {
+		return fmt.Errorf("required config param %s not found", configKeyASGName)
+	}
+	ctx := context.Background()
+
+	// Describe the ASG. This serves to both validate the config value is
+	// correct and ensure the AWS client is configured correctly. The response
+	// can also be used when performing the scaling, meaning we only need to
+	// call it once.
+	curASG, err := t.describeASG(ctx, asgName)
+	if err != nil {
+		return fmt.Errorf("failed to describe AWS Autoscaling Group: %v", err)
+	}
+
+	// The AWS ASG target requires different details depending on which
+	// direction we want to scale. Therefore calculate the direction and the
+	// relevant number so we can correctly perform the AWS work.
+	num, direction := t.calculateDirection(*curASG.DesiredCapacity, action.Count)
+
+	switch direction {
+	case "in":
+		err = t.scaleIn(ctx, curASG, num, config)
+	case "out":
+		err = t.scaleOut(ctx, curASG, num)
+	default:
+		return fmt.Errorf("scaling not required, ASG count %v and Autoscaler desired count %v",
+			*curASG.DesiredCapacity, action.Count)
```

Comment:
I don't feel strongly. I originally went with error because it means what the strategy calculated has been invalidated and might be helpful to know and raise. Logging this an returning nil seems adequate.